apps/crawler/lib/crawler/provider_behaviour.ex

```diff
@@ -0,0 +1,37 @@
+defmodule Crawler.Provider do
+  @moduledoc """
+  behaviour to implemented a provider.
+  `use` this module to implement the callbacks the standard way.
+  """
+  alias Crawler.Client
+
+  @doc "do a remote search, may return a error tuple"
+  @callback search(Client.query()) :: Client.search_result()
+  @doc "do a remote search, throws on error"
+  @callback search!(Client.query()) :: [Client.result()]
+
+  defmacro __using__(args) do
+    name = args[:name]
```

Comment:
pattern match in the function definition rather than parsing args. that way you get a compile-time error if you forget the name, not a run-time error

Reply:
args in __using__ is a keyword list.
pattern matching against a list, assumes the order of the arguments.
with just one possible option it is fine, but not really flexibile.

Reply:
how can we make this a compile-time error then? i can see this becoming a problem later on

Reply:
i guess the other option is to derive a default value somehow from the `__MODULE__`

Reply:
yeah, I tried and failed

Reply:
I could be wrong, btw.
Maybe `__using__` does not enforce `args` datatype.
getting `use Client, __MODULE__` would mean calling a macro, to call a macro.

Reply:
i'm gonna merge this as-is. we can find a fix for this later. i'll open an issue for each pending change in this pr.